multistate_kernel/kernel.py

```diff
@@ -211,27 +214,28 @@ def __call__(self, X, Y=None, eval_gradient=False):
 
         scale_index = np.zeros(shape=scale.shape, dtype=int)
         scale_index[np.tril_indices_from(scale_index)] = np.arange(len(np.tril_indices_from(scale_index)[0]))
-        scale_pos = 0
-        for kernel in self.state_kernels:
-            scale_pos += kernel.n_dims
+        scale_pos = sum(kernel.n_dims for kernel in self.state_kernels)
 
         kernel_value = np.zeros(shape=(n_samples_X, n_samples_Y, n_states))
         if eval_gradient:
-            kernel_gradient = np.zeros(shape=(n_samples_X, n_samples_Y, scale_pos))
+            kernel_gradient = np.zeros(shape=(n_samples_X, n_samples_X, scale_pos))
 
         pos = 0
         for k in range(n_states):
             kernel = self.state_kernels[k]
             if eval_gradient:
                 npos = pos + kernel.n_dims
-                kernel_value[...,k], kernel_gradient[...,range(pos,npos)] = kernel(X[:,1:], None, True)
+                kernel_value[...,k], kernel_gradient[...,pos:npos] = kernel(X[:,1:], None, True)
                 pos = npos
             else:
-                kernel_value[...,k] = kernel(X[:,1:], Y[:,1:], False)
+                if orig_Y is None:
+                    kernel_value[..., k] = kernel(X[:, 1:], None, False)
+                else:
+                    kernel_value[...,k] = kernel(X[:,1:], Y[:,1:], False)
```

Comment:
I see that this should be rebased ontop of the master. This is fixed via `X is Y` approach.